### PR TITLE
Remove OperationBindingContext

### PR DESCRIPTION
**Pull request by rynowak**
This change trims a few concepts that aren't really needed inside
ModelBinders anymore, and removes the OperationBindingContext class.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -178,18 +178,22 @@ internal static void AddMvcCoreServices(IServiceCollection services)
...
+            services.TryAdd(ServiceDescriptor.Transient<ICompositeMetadataDetailsProvider>(s =>
             {
-                var options = serviceProvider.GetRequiredService<IOptions<MvcOptions>>().Value;
+                var options = s.GetRequiredService<IOptions<MvcOptions>>().Value;
                 return new DefaultCompositeMetadataDetailsProvider(options.ModelMetadataDetailsProviders);
             }));
             services.TryAddSingleton<IModelBinderFactory, ModelBinderFactory>();
-            services.TryAddSingleton<IObjectModelValidator, DefaultObjectValidator>();
-            services.TryAddSingleton<ValidatorCache>();
+            services.TryAdd(ServiceDescriptor.Singleton<IObjectModelValidator>(s =>
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs]: Why not use the overload TryAddSingleton<T>(factory)?

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerContext.cs [hunk 2]
```diff
@@ -56,56 +52,6 @@ public new ControllerActionDescriptor ActionDescriptor
...
-                }
-
-                _inputFormatters = value;
-            }
-        }
-
-        /// <summary>
-        /// Gets or sets the list of <see cref="IModelValidatorProvider"/> instances for the current request.
-        /// </summary>
-        public virtual IList<IModelValidatorProvider> ValidatorProviders
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerContext.cs]: Wow, we were carting these two properties around for `TryUpdateModelAsync()` calls and that's it? No real scenario for controller code to change `ValidatorProviders` before calling `TryValidateModel()`?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs [hunk 3]
```diff
@@ -39,20 +39,20 @@ public class ControllerArgumentBinder : IControllerActionArgumentBinder
         }
 
         public Task<IDictionary<string, object>> BindActionArgumentsAsync(
-            ControllerContext context,
+            ControllerContext controllerContext,
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs]: Renaming the parameters makes it harder to follow along at home.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs [hunk 4]
```diff
@@ -27,22 +30,24 @@ public DefaultModelBindingContext()
...
             string modelName)
         {
-            if (operationBindingContext == null)
+            if (actionContext == null)
             {
-                throw new ArgumentNullException(nameof(operationBindingContext));
+                throw new ArgumentNullException(nameof(actionContext));
             }
 
             if (metadata == null)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: Why is `valueProvider == null` allowed?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs [hunk 5]
```diff
@@ -113,7 +115,7 @@ public DefaultModelBindingContext()
             // to preserve the currrent state.
             if (modelMetadata.BindingSource != null && !modelMetadata.BindingSource.IsGreedy)
             {
-                ValueProvider = FilterValueProvider(_operationBindingContext.ValueProvider, modelMetadata.BindingSource);
+                ValueProvider = FilterValueProvider(_originalValueProvider, modelMetadata.BindingSource);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: Why isn't this using the `OriginalValueProvider` property?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs [hunk 6]
```diff
@@ -130,32 +132,16 @@ public DefaultModelBindingContext()
         }
 
         /// <inheritdoc />
-        public override NestedScope EnterNestedScope()
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: Let's do a pass to reorder this class _separately_. This move doesn't do much more than make the review harder to read.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: I moved about 10 lines of code to comply with our coding standards. If I hadn't moved it, there's a good chance you might have suggested I move it :see_no_evil:

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: 1. You split these methods from the other `EnterNestedScope()` overload.
2. There is no such guideline; no mention of member order in https://github.com/aspnet/Home/wiki/Engineering-guidelines#coding-guidelines. It's an informal policy we've imposed on new code and not something we rewrite classes to match.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs [hunk 7]
```diff
@@ -270,8 +269,15 @@ public override IValueProvider ValueProvider
         /// <inheritdoc />
         public override ValidationStateDictionary ValidationState
         {
-            get { return _state.ValidationState; }
-            set { _state.ValidationState = value; }
+            get { return _validationState; }
+            set
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: Looks like this setter exists only for tests, when used together with the test-only parameterless constructor. Please add comments because it's too easy to mess the non-stacked property values up if the assumptions aren't stated.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs [hunk 8]
```diff
@@ -14,34 +15,36 @@ public class DefaultObjectValidator : IObjectModelValidator
     {
         private readonly IModelMetadataProvider _modelMetadataProvider;
         private readonly ValidatorCache _validatorCache;
+        private readonly IModelValidatorProvider _validatorProvider;
 
         /// <summary>
         /// Initializes a new instance of <see cref="DefaultObjectValidator"/>.
         /// </summary>
         /// <param name="modelMetadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
-        /// <param name="validatorCache">The <see cref="ValidatorCache"/>.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs]: Strange to take the `ValidatorCache` out of DI. Why does the `DefaultObjectValidator` now own the instance?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs [hunk 9]
```diff
@@ -51,14 +54,14 @@ public class DefaultObjectValidator : IObjectModelValidator
                 throw new ArgumentNullException(nameof(actionContext));
             }
 
-            if (validatorProvider == null)
+            if (validationState == null)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs]: The `ValidationVisitor` handles `validationState == null` just fine. Why does this condition now lead to an `ArgumentNullException`?

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs [hunk 10]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.IO;
 using System.Linq;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs]: Can this `using` be removed now?

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/ComplexTypeModelBinder.cs [hunk 11]
```diff
@@ -271,33 +271,6 @@ private bool CanValueBindAnyModelProperties(ModelBindingContext bindingContext)
...
-                if (rootValueProvider != null)
-                {
-                    valueProvider = rootValueProvider.Filter(bindingSource);
-                    if (valueProvider == null)
-                    {
-                        // Unable to find a value provider for this binding source. Binding will fail.
-                        return false;
-                    }
-                }
-            }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/ComplexTypeModelBinder.cs]: Where did this block move to? Or, are you just no longer worried about the eventual binding failure?

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/DictionaryModelBinder.cs [hunk 12]
```diff
@@ -81,8 +81,8 @@ public DictionaryModelBinder(IModelBinder keyBinder, IModelBinder valueBinder)
             }
 
             // Update the existing successful but empty ModelBindingResult.
-            var metadataProvider = bindingContext.OperationBindingContext.MetadataProvider;
-            var valueMetadata = metadataProvider.GetMetadataForType(typeof(TValue));
+            var elementMetadata = bindingContext.ModelMetadata.ElementMetadata;
+            var valueMetadata = elementMetadata.Properties[nameof(KeyValuePair<TKey, TValue>.Value)];
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/DictionaryModelBinder.cs]: Nice.

File: src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/ApiController.cs [hunk 13]
```diff
@@ -567,11 +567,10 @@ public void Validate<TEntity>(TEntity entity)
         /// </param>
         public void Validate<TEntity>(TEntity entity, string keyPrefix)
         {
-            var validatidationState = new ValidationStateDictionary();
+            var validationState = new ValidationStateDictionary();
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/ApiController.cs]: No, no. I'm positive we meant to introduce the new concept of `validatidation` states :trollface:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/DefaultObjectValidatorTests.cs [hunk 14]
```diff
@@ -466,12 +453,7 @@ public void Validate_ComplexType_IValidatableObject_CanUseRequestServices()
                 .Returns(new List<ValidationResult>());
 
             // Act
-            validator.Validate(
-                actionContext, 
-                validatorProvider, 
-                validationState, 
-                null, 
-                model.Object);
+            validator.Validate(actionContext, validationState, null, model.Object);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/DefaultObjectValidatorTests.cs]: minor: plz name the `null` guy

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs [hunk 15]
```diff
@@ -125,7 +125,7 @@ public class KeyValuePairModelBinderTest
             var binder = new KeyValuePairModelBinder<int, string>(innerBinder, innerBinder);
 
             // Act
-            var result = await binder.TryBindStrongModel<int>(bindingContext, innerBinder, "key");
+            var result = await binder.TryBindStrongModel<int>(bindingContext, innerBinder, "Key");
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs]: Ah, previously case didn't matter but it does now. Correct?

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs [hunk 16]
```diff
@@ -212,21 +208,13 @@ private static DefaultModelBindingContext CreateContext()
             IValueProvider valueProvider,
             Type keyValuePairType)
         {
-            var metataProvider = new EmptyModelMetadataProvider();
+            var metadataProvider = new TestModelMetadataProvider();
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs]: Does changing from an `EmptyModelMetadataProvider` to a `TestModelMetadataProvider` mean anything?

File: test/Microsoft.AspNetCore.Mvc.IntegrationTests/ModelBindingTestContext.cs [hunk 17]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc.IntegrationTests
+{
+    public class ModelBindingTestContext : ControllerContext
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.IntegrationTests/ModelBindingTestContext.cs]: Like it! Was getting concerned about reanimating `OperationBindingContext` as a test concept. But this is a nice hedge to reduce test code churn.

**Comment by dougbu** [comment]
Suggest we'll need an announcement for those (few?) users that call `ModelBindingHelper` methods directly or who have written their own model binders.

**Comment by dougbu** [comment]
:watch: for questions and lower-level comments. Liking the overall direction i.e. the boilerplate deletions.

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerContext.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerContext.cs]: Correct, and yes, no real scenario.

File: src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs]: Sure

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: fixed

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: Fixed

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs (hunk 7 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: Added some comments. Really **all** of these are unit testing only btw.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs (hunk 8 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs]: Because it would never be used anywhere else, and it can just construct the cache. Putting the cache in DI just adds complexity

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs]: Sure, will revert this change.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs (hunk 10 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs]: yep

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/ComplexTypeModelBinder.cs (hunk 11 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/ComplexTypeModelBinder.cs]: This moved inside the binding context. I'm not worried about it because you'll fail to find any values in an empty collection

File: src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/ApiController.cs (hunk 13 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.WebApiCompatShim/ApiController.cs]: :heart_eyes:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs (hunk 16 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs]: Not really, I was just copy-pasting things

**Comment by rynowak** [comment]
@dougbu updated

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs [hunk 18]
```diff
@@ -1429,8 +1415,7 @@ public virtual SignOutResult SignOut(AuthenticationProperties properties, params
 
             ObjectValidator.Validate(
                 ControllerContext,
-                new CompositeModelValidatorProvider(ControllerContext.ValidatorProviders),
-                validationState: null,
+                validationState: null
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: Won't compile unless you restore the comma.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs (hunk 15 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs]: Unanswered. Am I missing anything here?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs (hunk 7 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: Sounds good.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs (hunk 6 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultModelBindingContext.cs]: This wasn't an important part of this PR but you haven't actually fixed anything yet. Either revert this part of the change or move `CreateBindingContext()` and the other `EnterNestedScope()` after the properties as well.

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/ComplexTypeModelBinder.cs (hunk 11 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/ComplexTypeModelBinder.cs]: :ok:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs (hunk 8 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultObjectValidator.cs]: :ok:

**Comment by dougbu** [comment]
:shipit: with a couple of fixes (especially [this](https://github.com/aspnet/Mvc/pull/4501#discussion-diff-60352310) and [this](https://github.com/aspnet/Mvc/pull/4501#discussion-diff-60858171)) -- as long as I'm not missing something [here](https://github.com/aspnet/Mvc/pull/4501#discussion-diff-60447549).

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs (hunk 15 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/KeyValuePairModelBinderTest.cs]: Yeah sorry missed this one. There's no trick. `.Properties[string]` is case-sensitive (as all things should be)
